statsmodels/tsa/tsatools.py

```diff
@@ -36,54 +37,62 @@ def add_trend(X, trend="c", prepend=False, has_constant='skip'):
     --------
     statsmodels.add_constant
     """
-    #TODO: could be generalized for trend of aribitrary order
+    # TODO: could be generalized for trend of aribitrary order
     trend = trend.lower()
-    if trend == "c":    # handles structured arrays
+    columns = ['const', 'trend', 'trend_squared']
+    if trend == "c":  # handles structured arrays
         return add_constant(X, prepend=prepend, has_constant=has_constant)
     elif trend == "ct" or trend == "t":
+        columns = columns[:2]
+        if trend == "t":
+            columns = columns[1:2]
         trendorder = 1
     elif trend == "ctt":
         trendorder = 2
     else:
         raise ValueError("trend %s not understood" % trend)
 
-    X = np.asanyarray(X)
+    is_recarray = _is_recarray(X)
+    is_pandas = _is_using_pandas(X, None) or is_recarray
+    if is_pandas or is_recarray:
+        import pandas as pd
+
+        if is_recarray:
+            X = pd.DataFrame.from_records(X)
+        elif isinstance(X, pd.Series):
+            X = pd.DataFrame(X)
+        else:
+            X = X.copy()
+    else:
+        X = np.asanyarray(X)
+
     nobs = len(X)
-    trendarr = np.vander(np.arange(1,nobs+1, dtype=float), trendorder+1)
+    trendarr = np.vander(np.arange(1, nobs + 1, dtype=np.float64), trendorder + 1)
     # put in order ctt
     trendarr = np.fliplr(trendarr)
     if trend == "t":
-        trendarr = trendarr[:,1]
-    if not X.dtype.names:
-        # check for constant
-        if "c" in trend and np.any(np.ptp(X, axis=0) == 0):
-            if has_constant == 'raise':
-                raise ValueError("X already contains a constant")
-            elif has_constant == 'add':
-                pass
-            elif has_constant == 'skip' and trend == "ct":
-                trendarr = trendarr[:, 1]
-
-        if not prepend:
-            X = np.column_stack((X, trendarr))
-        else:
-            X = np.column_stack((trendarr, X))
+        trendarr = trendarr[:, 1]
+
+    if "c" in trend and np.any(np.ptp(np.asanyarray(X), axis=0) == 0):
+        if has_constant == 'raise':
+            raise ValueError("X already contains a constant")
+        elif has_constant == 'add':
+            pass
+        elif has_constant == 'skip' and trend == "ct":
```

Comment:
this need to handle also "ctt", i.e. `trend in ["ct", "ctt"]` or  `trend_order > 0`